include/LightGBM/dataset.h

```diff
@@ -495,10 +495,21 @@ class Dataset {
 
   inline void set_feature_names(const std::vector<std::string>& feature_names) {
     if (feature_names.size() != static_cast<size_t>(num_total_features_)) {
-      Log::Warning("size of feature_names error, should equal with total number of features");
+      Log::Warning("Size of feature_names error, should equal with total number of features");
```

Comment:
I changed `dataset->feature_names_ = feature_names_;` to `dataset->set_feature_names(feature_names_);`. so if `feature_names.size() != static_cast<size_t>(num_total_features_)`, `dataset->feature_names_` will be null. I didn't change this yet, you can review rest of pr first.